Win32-network/src/System/Win32/Async/Socket.hs

```diff
@@ -35,16 +34,18 @@ sendBuf :: Socket
 sendBuf sock buf size = Socket.withFdSocket sock $ \fd ->
     -- on Windows sockets are Word32, GHC represents file descriptors with CInt
     -- which is Int32.
-    alloca $ \bufs_ptr -> do
-      poke bufs_ptr WSABuf {buf, len = fromIntegral size}
-      wsaWaitForCompletion "sendBuf" (c_sendBuf fd bufs_ptr 1)
-
-foreign import ccall safe "HsSendBuf"
-    c_sendBuf :: SOCKET
-              -> Ptr WSABuf  -- ^ lpBuffers
-              -> DWORD       -- ^ dwBufferCount
-              -> StablePtr b
-              -> IO ()
+    alloca $ \bufsPtr ->
+      withIOCPData "sendBuf" (FDSocket fd) $ \lpOverlapped waitVar -> do
+        poke bufsPtr WSABuf {buf, len = fromIntegral size}
+        sendResult <- c_WSASend fd bufsPtr 1 nullPtr 0 lpOverlapped nullPtr
+        errorCode <- wsaGetLastError
+        if sendResult == 0 || errorCode == wSA_IO_PENDING
```

Comment:
same Q as elsehwhere about whether this is the simplest condition.

Reply:
I believe so.  Tracing these values prints two cases (the first one is `sendResult` the other one is `errorCode` - `997`, except being the police telephone number in Poland :D, is `WSA_IO_PENDING`) :
```
recvBuf -1, 997
recvBuf -1, 997
recvBuf -1, 997
recvBuf -1, 997
recvBuf 0, 0
recvBuf 0, 0
recvBuf -1, 997

```

Reply:
^ this output was generated for `recvBuf` rather than `sendBuf`; for `sendBuf` it's similar (just `0, 0` is much more common):
```
sendBuf 0, 0
sendBuf 0, 0
sendBuf 0, 0
sendBuf 0, 0
sendBuf 0, 0
sendBuf -1, 997
sendBuf 0, 0
sendBuf 0, 0
sendBuf 0, 0
sendBuf 0, 0
```

Reply:
Ok but in these examples the result being zero or non-zero always agrees with the error code. So we would be able to base our condition on either of them rather than both.

This is not a blocker, and I suggest we merge without addressing this now, but it certainly looks odd and we could clean this up later. File a low priority ticket.